app/src/main/java/com/example/drawingactivity/DialogUtils.kt

```diff
@@ -137,6 +136,183 @@ object DialogUtils {
+        val greenText = view.findViewById<ForcableEditText>(R.id.greenText)
+        val blueText = view.findViewById<ForcableEditText>(R.id.blueText)
+        val topColorSquare = view.findViewById<View>(R.id.topColor)
+        val bottomColorSquare = view.findViewById<View>(R.id.bottomColor)
+        val confirmButton = view.findViewById<Button>(R.id.okButton)
+        confirmButton.setOnClickListener {
+            val submitArray = FloatArray(3)
+            submitArray[0] = (hueText.text.toString().toFloat())
+            submitArray[1] = (satText.text.toString().toFloat() / 100)
+            submitArray[2] = (valText.text.toString().toFloat() / 100)
+            onSubmit.invoke(submitArray, hexText.text.toString())
+            sheet.dismiss()
+        }
+        val cancelButton = view.findViewById<Button>(R.id.nopeButton)
+        cancelButton.setOnClickListener {
+            onSubmit.invoke(null, null)
+            sheet.dismiss()
+        }
+        sheet.setOnCancelListener {
+            onSubmit.invoke(null, null)
+            sheet.dismiss()
+        }
+        /**
+         * Function that updates all the views in the gradient.
+         * Takes in a color state data class to update views.
+         */
+        fun forceViewUpdates(colorState: ColorUtils.ColorState) {
+            val displaySat = (colorState.saturation * 100).toInt()
+            val displayVal = (colorState.value * 100).toInt()
+            hexText.forceText(colorState.hexCode)
+            hueText.forceText(colorState.hue.toString())
+            satText.forceText(displaySat.toString())
+            valText.forceText(displayVal.toString())
+            redText.forceText(colorState.red.toString())
+            greenText.forceText(colorState.green.toString())
+            blueText.forceText(colorState.blue.toString())
+            huePicker.setHue(colorState.hue)
+            gradientPicker.setHue(colorState.hue)
+            gradientPicker.setSaturationAndValue(colorState.saturation, colorState.value)
+            topColorSquare.setBackgroundColor(parseColor(colorState.hexCode))
+        }
+
+        /**
+         * Clears focus on whatever view the user has currently selected.
+         * Currently only called when the user interacts with the Gradient or Hue Slider.
+         */
+        fun clearFocus() {
+            hexText.clearFocus()
+            hueText.clearFocus()
+            satText.clearFocus()
+            valText.clearFocus()
+            redText.clearFocus()
+            greenText.clearFocus()
+            blueText.clearFocus()
+        }
+        val colorStates = ColorUtils.calculateColorWithHSV(previousColor)
+        forceViewUpdates(colorStates)
+        bottomColorSquare.setBackgroundColor(parseColor(colorStates.hexCode))
+        huePicker?.onHueChanged = {
+            clearFocus()
+            val hsv = FloatArray(3)
+            hsv[0] = it.toFloat()
+            hsv[1] = (satText.text.toString().toFloat() / 100)
```

Comment:
shouldn't this be toFloatorNull? what if they give you a fucky value?

Reply:
See below, this is the same problem described there and I need to add in validation on my checks to prevent crashes.